### PR TITLE
Adding 3D geometrical tools

### PR DESCRIPTION
**Pull request by icedcoffeeee**
## Changelog / Overview

- Added new :class:`~.Plane` mobject that takes in a point and a normal vector
- Added two static methods for :class:`~.Line3D` that allows a line to made relatively perpendicular or parallel to another line.
- Added multiple functions in `space_ops.py`
(:func:`~.projected_vector`, :func:`~.cartesian_to_spherical`)

## Motivation

An attempt to implement some of the features of issue #1789

## Explanation for Changes

[Plane Documentation](https://manimce--1798.org.readthedocs.build/en/1798/reference/manim.mobject.three_dimensions.Plane.html)
[Line3D Documentation](https://manimce--1798.org.readthedocs.build/en/1798/reference/manim.mobject.three_dimensions.Line3D.html)

## Testing Status

All tests passed locally
Scenes used:
```py
class PlaneTest(ThreeDScene):
    def construct(self):
        self.set_camera_orientation(PI/3, 100*DEGREES)
        self.add(ax := ThreeDAxes())
        normal = [1,1,1]
        self.add(Arrow3D(ORIGIN,normal))
        self.add(Plane(ORIGIN,normal))
        self.wait()
```
```py
class PerpendicularLineTest(ThreeDScene):
    def construct(self):
        self.set_camera_orientation(PI / 3, -PI / 3)
        self.add(ax := ThreeDAxes())
        self.add(a := Line3D([-1, -1, -1], [1, 1, 1], color=RED))
        self.add(b := Dot3D([-1, 1, 2], color=BLUE))
        self.add(c := Line3D.perp_to_line(b.get_center(), a, color=GREEN))
        self.wait()
```
## Further Comments

The other features mentioned seems too complicated for my level of expertise😅 
## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [x] My new functions/classes either have a docstring or are private
- [x] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

**Comment by torannn** [comment]
Awesome, really appreciate your effort. I don't know much about coding in python but I will try adding some algorithms to build such things.

**Comment by icedcoffeeee** [comment]
@torannn Thanks. Everything is finished, except for the fact I can't seem to figure out why the test are failing 🤔 I'll check it tomorrow

File: manim/mobject/three_dimensions.py [hunk 1]
```diff
@@ -732,6 +791,71 @@ def get_start(self):
     def get_end(self):
         return self.end
 
+    @staticmethod
+    def perp_to_line(point, line, **kwargs):
```
Comment (hydrobeam) [inline manim/mobject/three_dimensions.py:795]: An implementation with a `classmethod` would be more appropriate here. Plus, being able to control the length of the line would be nice.

File: manim/mobject/three_dimensions.py [hunk 2]
```diff
@@ -732,6 +791,71 @@ def get_start(self):
...
+            The line to be perpendicular to.
+        """
+        u = point - line.get_center()
+        v = line.vect
+        point2 = projected_vector(v, u)
+        direction = point - point2
+        return Line3D(point2 + direction * 10, point2 - direction * 10, **kwargs)
+
+    @staticmethod
+    def parallel_to_line(point, line, **kwargs):
```
Comment (hydrobeam) [inline manim/mobject/three_dimensions.py:828]: same as above

File: manim/mobject/three_dimensions.py [hunk 3]
```diff
@@ -732,6 +791,71 @@ def get_start(self):
...
+        point: :class:`numpy.ndarray`
+            A numpy array of the point at which the line passes
+        line: :class:`Line3D`
+            The line to be perpendicular to.
+        """
+        u = point - line.get_center()
+        v = line.vect
+        point2 = projected_vector(v, u)
+        direction = point - point2
+        return Line3D(point2 + direction * 10, point2 - direction * 10, **kwargs)
```
Comment (hydrobeam) [inline manim/mobject/three_dimensions.py:825]: It'd probably make more sense if the generated line was the same length as the line that it's based on.

File: manim/mobject/three_dimensions.py [hunk 4]
```diff
@@ -732,6 +791,71 @@ def get_start(self):
...
+        ----------
+        point: :class:`numpy.ndarray`
+            A numpy array of the point at which the line passes
+        line: :class:`Line3D`
+            The line to be parallel to.
+        """
+        u = point - line.get_center()
+        v = line.vect
+        direction = projected_vector(v, u)
+        return Line3D(point + direction * 10, point - direction * 10, **kwargs)
```
Comment (hydrobeam) [inline manim/mobject/three_dimensions.py:857]: same as above

File: manim/mobject/three_dimensions.py [hunk 5]
```diff
@@ -254,6 +260,59 @@ def param_surface(u, v):
...
+    ----------
+    point: :class:`numpy.ndarray`
+        A point that lies on the plane.
+    normal_vect: :class:`numpy.ndarray`
+        A vector perpendicular to the plane.
+    kwargs
+        To be passed to :class:`ParametricSurface`
+    """
+
+    def __init__(self, point=ORIGIN, normal_vect=[1, 1, 1], **kwargs):
```
Comment (hydrobeam) [inline manim/mobject/three_dimensions.py:294]: the typing should be brought into the `__init__` here.

File: manim/mobject/three_dimensions.py [hunk 6]
```diff
@@ -254,6 +260,59 @@ def param_surface(u, v):
...
+        v_range = kwargs.pop("v_range", [-2 * PI, 2 * PI])
+        super().__init__(
+            lambda u, v: np.array([u, v, 0]), u_range=u_range, v_range=v_range, **kwargs
+        )
+        spherical = cartesian_to_spherical(normal_vect)
+        axis = [np.sin(spherical[1]), -np.cos(spherical[1]), 0]
+        self.rotate(-spherical[2], axis=axis)
+        self.shift(projected_vector(self.normal_vect, point))
+
+    def line_perp_to_plane(self, point, **kwargs):
```
Comment (hydrobeam) [inline manim/mobject/three_dimensions.py:306]: typing + a return type (can be specifed by adding a `->` before the colon), Plus, I think the length of the line should be able to be adjusted here.

File: manim/mobject/three_dimensions.py [hunk 7]
```diff
@@ -732,6 +791,71 @@ def get_start(self):
...
+                    self.set_camera_orientation(45 * DEGREES, 135 * DEGREES)
+                    ax = ThreeDAxes()
+                    line1 = Line3D([-5, -5, -5], [5, 5, 5], color=RED)
+                    dot = Dot3D([3, 0, 2], color=BLUE)
+                    line2 = Line3D.perp_to_line(dot.get_center(), line1, color=YELLOW)
+                    self.add(ax, line1, dot, line2)
+
+        Parameters
+        ----------
+        point: :class:`numpy.ndarray`
```
Comment (hydrobeam) [inline manim/mobject/three_dimensions.py:816]:
```suggestion
        point: Sequence[float]
```

File: manim/utils/space_ops.py [hunk 8]
```diff
@@ -775,3 +778,67 @@ def perpendicular_bisector(
...
+    numpy array
+        The component vector of the second vector
+        in the direction of the first vector
+    """
+    dot_prod = np.dot(vec1, vec2)
+    length_prod = np.linalg.norm(vec1) ** 2
+    return vec1 * dot_prod / length_prod
+
+
+def cartesian_to_spherical(point: Union[list, np.ndarray]):
```
Comment (hydrobeam) [inline manim/utils/space_ops.py:804]: the return type is an `np.ndarray` here
```suggestion
def cartesian_to_spherical(point: Sequence[float]):
```

File: manim/utils/space_ops.py [hunk 9]
```diff
@@ -775,3 +778,67 @@ def perpendicular_bisector(
...
+        A list of numbers corresponding to each
+        polar coordinate value (rho, theta, phi)
+    """
+    phi = angle_between_vectors(point, OUT)
+    theta = np.arctan2(point[1], point[0])
+    rho = np.linalg.norm(point)
+    return np.array([rho, theta, phi])
+
+
+def spherical_to_cartesian(spherical: Union[list, np.ndarray]):
```
Comment (hydrobeam) [inline manim/utils/space_ops.py:825]:
```suggestion
def spherical_to_cartesian(spherical: Sequence[float]):
```

File: manim/mobject/three_dimensions.py [hunk 10]
```diff
@@ -732,6 +791,71 @@ def get_start(self):
...
+            :save_last_frame:
+
+            class PerpendicularLineExample(ThreeDScene):
+                def construct(self):
+                    self.set_camera_orientation(45 * DEGREES, 135 * DEGREES)
+                    ax = ThreeDAxes()
+                    line1 = Line3D([-5, -5, -5], [5, 5, 5], color=RED)
+                    dot = Dot3D([3, 0, 2], color=BLUE)
+                    line2 = Line3D.perp_to_line(dot.get_center(), line1, color=YELLOW)
+                    self.add(ax, line1, dot, line2)
```
Comment (hydrobeam) [inline manim/mobject/three_dimensions.py:812]: The angle feels a little too high for this example + its sister method. Maybe bringing to a more central position would look nicer. Also, using `.c2p` to make the lines based on the axes (right now it does nothing 🤔) might also be wise.

image

**Review by hydrobeam — CHANGES_REQUESTED** [review]
Left some typing comments + comments about the sister methods being made more extendable. Typing should be done in the `__init__` and also be sure to specify return types on the methods using `->`.

File: manim/mobject/three_dimensions.py [hunk 11]
```diff
@@ -254,6 +260,59 @@ def param_surface(u, v):
...
+        class PlaneExample(ThreeDScene):
+            def construct(self):
+                self.set_camera_orientation(PI / 3, 100 * DEGREES)
+                ax = ThreeDAxes()
+                normal = [1, 1, 1]
+                normal_vector = Arrow3D(ORIGIN, normal, color=RED)
+                dot = Dot3D([-2, -1, 1], color=ORANGE)
+                plane = Plane(point=dot.get_center(), normal_vect=normal)
+                self.add(ax, normal_vector, dot, plane)
+                self.wait()
```
[the remark this answers is not in the document]
Comment (Darylgolden) [reply manim/mobject/three_dimensions.py:282]: Yeah, perhaps decrease the opacity.

[the remark this answers is not in the document]
Comment (icedcoffeeee) [reply manim/mobject/three_dimensions.py:282]: PlaneExample_ManimCE_v0 7 0
Would something like this be nicer? 🤔

File: manim/mobject/three_dimensions.py [hunk 12]
```diff
@@ -254,6 +260,106 @@ def param_surface(u, v):
...
+                    v_range=[-2.5, 2.5],
+                    fill_opacity=0.5,
+                )
+                normal_vector.shift(plane.get_center())
+                self.add(ax, normal_vector, dot, plane)
+                self.wait()
+
+    Parameters
+    ----------
+    point: Sequence[float]
```
Comment (PhotonSpheres) [inline manim/mobject/three_dimensions.py:293]:
```suggestion
    point
```
No need for type hints here anymore.

File: manim/mobject/three_dimensions.py [hunk 13]
```diff
@@ -254,6 +260,106 @@ def param_surface(u, v):
...
+        A point that lies on the plane.
+    normal_vect: Sequence[float]
```
Comment (PhotonSpheres) [inline manim/mobject/three_dimensions.py:295]:
```suggestion
    normal_vect
```

File: manim/mobject/three_dimensions.py [hunk 14]
```diff
@@ -254,6 +260,106 @@ def param_surface(u, v):
...
+        A vector perpendicular to the plane.
+    u_range: Sequence[float]
```
Comment (PhotonSpheres) [inline manim/mobject/three_dimensions.py:297]:
```suggestion
    u_range
```

File: manim/mobject/three_dimensions.py [hunk 15]
```diff
@@ -254,6 +260,106 @@ def param_surface(u, v):
 # Specific shapes
 
 
+class Plane(ParametricSurface):
+    """
+    A 3D Plane Mobject.
```
Comment (PhotonSpheres) [inline manim/mobject/three_dimensions.py:265]:
```suggestion
    A 2D plane :class:`~.VMobject`.
```

File: manim/mobject/three_dimensions.py [hunk 16]
```diff
@@ -254,6 +260,106 @@ def param_surface(u, v):
...
+
+    Parameters
+    ----------
+    point: Sequence[float]
+        A point that lies on the plane.
+    normal_vect: Sequence[float]
+        A vector perpendicular to the plane.
+    u_range: Sequence[float]
+        The dimensions of the plane.
+    v_range: Sequence[float]
```
Comment (PhotonSpheres) [inline manim/mobject/three_dimensions.py:299]:
```suggestion
    v_range
```

File: manim/mobject/three_dimensions.py [hunk 17]
```diff
@@ -254,6 +260,106 @@ def param_surface(u, v):
...
+        The dimensions of the plane.
+    kwargs
+        To be passed to :class:`ParametricSurface`
```
Comment (PhotonSpheres) [inline manim/mobject/three_dimensions.py:302]:
```suggestion
        To be passed to :class:`ParametricSurface`.
```

File: manim/utils/space_ops.py [hunk 18]
```diff
@@ -775,3 +778,67 @@ def perpendicular_bisector(
     direction = np.cross(p1 - p2, norm_vector)
     m = midpoint(p1, p2)
     return [m + direction, m - direction]
+
+
+def projected_vector(vec1: np.ndarray, vec2: np.ndarray):
```
Comment (PhotonSpheres) [inline manim/utils/space_ops.py:783]:
```suggestion
def projected_vector(vec1: np.ndarray, vec2: np.ndarray) -> np.ndarray:
```

File: manim/utils/space_ops.py [hunk 19]
```diff
@@ -775,3 +778,67 @@ def perpendicular_bisector(
...
+    numpy array
+        The component vector of the second vector
+        in the direction of the first vector
+    """
+    dot_prod = np.dot(vec1, vec2)
+    length_prod = np.linalg.norm(vec1) ** 2
+    return vec1 * dot_prod / length_prod
+
+
+def cartesian_to_spherical(point: Sequence[float]):
```
Comment (PhotonSpheres) [inline manim/utils/space_ops.py:804]:
```suggestion
def cartesian_to_spherical(point: Sequence[float]) -> np.ndarray:
```

File: manim/utils/space_ops.py [hunk 20]
```diff
@@ -775,3 +778,67 @@ def perpendicular_bisector(
...
+    """Returns a list of numbers corresponding to each
+    polar coordinate value (phi, theta, distance).
+
+    Parameters
+    ----------
+    point
+        a numpy array point.
```
Comment (PhotonSpheres) [inline manim/utils/space_ops.py:811]:
```suggestion
        A numpy array point.
```

File: manim/utils/space_ops.py [hunk 21]
```diff
@@ -775,3 +778,67 @@ def perpendicular_bisector(
...
+        The component vector of the second vector
+        in the direction of the first vector
+    """
+    dot_prod = np.dot(vec1, vec2)
+    length_prod = np.linalg.norm(vec1) ** 2
+    return vec1 * dot_prod / length_prod
+
+
+def cartesian_to_spherical(point: Sequence[float]):
+    """Returns a list of numbers corresponding to each
```
Comment (PhotonSpheres) [inline manim/utils/space_ops.py:805]:
```suggestion
    """Returns a numpy array of numbers corresponding to each
```

File: manim/utils/space_ops.py [hunk 22]
```diff
@@ -775,3 +778,67 @@ def perpendicular_bisector(
...
+    Parameters
+    ----------
+    point
+        a numpy array point.
+
+    Returns
+    -------
+    numpy array
+        A list of numbers corresponding to each
+        polar coordinate value (rho, theta, phi)
```
Comment (PhotonSpheres) [inline manim/utils/space_ops.py:817]:
```suggestion
        polar coordinate value (rho, theta, phi).
```

File: manim/utils/space_ops.py [hunk 23]
```diff
@@ -775,3 +778,67 @@ def perpendicular_bisector(
...
+    """
+    phi = angle_between_vectors(point, OUT)
+    theta = np.arctan2(point[1], point[0])
+    rho = np.linalg.norm(point)
+    return np.array([rho, theta, phi])
+
+
+def spherical_to_cartesian(spherical: Sequence[float]):
```
Comment (PhotonSpheres) [inline manim/utils/space_ops.py:825]:
```suggestion
def spherical_to_cartesian(spherical: Sequence[float]) -> np.ndarray:
```

File: manim/utils/space_ops.py [hunk 24]
```diff
@@ -775,3 +778,67 @@ def perpendicular_bisector(
...
+    """Returns a point coordinate based on the spherical
+    coordinates given.
+
+    Parameters
+    ----------
+    spherical
+        a numpy array of spherical coordinates.
```
Comment (PhotonSpheres) [inline manim/utils/space_ops.py:832]:
```suggestion
        A numpy array of spherical coordinates.
```

**Review by PhotonSpheres — COMMENTED** [review]
Just some type hints and docstrings remarks.

**Review by behackl — CHANGES_REQUESTED** [review]
Thanks to #1802 #1835 #1836 we have now a much better testing system -- the `PlaneTest` added with this PR should be rewritten accordingly.

It is very easy though, just compare how the other tests in `test_threed.py` look in the most recent main branch. The control data can then be renewed by running `pytest --set_test tests/test_graphical_units/test_threed.py:test_Plane` (or something very similar to that).

Something we should also be aware of: given that the implementation of `Plane` here inherits from `ParametricSurface`, this will also suffer from problems related to OpenGL rendering.

File: manim/mobject/three_dimensions.py [hunk 25]
```diff
@@ -732,6 +838,86 @@ def get_start(self):
...
+        return Line3D(
+            intersection + direction * length / 2,
+            intersection - direction * length / 2,
+            **kwargs,
+        )
+
+    @classmethod
+    def parallel_to_line(
+        cls, point: Sequence[float], line: "Line3D", length: float = None, **kwargs
+    ) -> "Line3D":
```
Comment (hydrobeam) [inline manim/mobject/three_dimensions.py:887]: The default of `None` breaks the code here, it's not implemented as it was in `perp_to_line`
```suggestion
    @classmethod
    def parallel_to_line(
        cls, point: Sequence[float], line: Line3D, length: float = None, **kwargs
    ) -> Line3D:
```

Comment (icedcoffeeee) [reply manim/mobject/three_dimensions.py:887]: I can fix the length attribute, but why should the quotation marks be dropped? it leaves a undefined error if it's dropped

**Comment by icedcoffeeee** [comment]
I just had the realization, it might be way better to just use a `Square()` instead 😗

**Comment by icedcoffeeee** [comment]
> use a `Square()` instead

nvm, that doesn't look right.

**Comment by icedcoffeeee** [comment]
What's the opinion on changing inheritance of the Plane from ParametricSurface to simply ThreeDVMobject?
I tried rendering the new one, and I kind of like it. For comparison:
Old:
PlaneExample-1
New:
Test_ManimCE_v0 7 0
Benefits:
- lower rendering time since resolution isn't a concern
- Reliable points. Using ParametricSurface sometimes returns no points? Also a problem with points not rotating in opengl.

Disadvantages:
- Haven't thought of any, hence my question.

Notes:
- The scaling seems off (the code is correct, not sure why)

**Review by Darylgolden — CHANGES_REQUESTED** [review]
Hi, sorry for taking so long to get back to you! While making the plane inherit from a Surface may not make much sense for Cairo (since a Surface there is just a group of ThreeDVmobjects), it does make sense for OpenGL. This also ensures consistent behavior when #1843 is merged. What do you think?

**Comment by icedcoffeeee** [comment]
I left this PR for quite a while, and looking at it now, I kind of see this PR as a bit ambitious. It's introducing a new mobject Plane, adding functions to Line3D,  as well as adding and modifying functions in space_ops.py. A lot of the codebase has also changed since this PR was first open too. 

That being said, I personally think I should close this PR, and start again with smaller PRs that bring these changes separately. Hopefully I can rewrite the changes to be more readable and efficient, as well as helping reviewers to review smaller amounts of code rather than a huge change as it currently is.

**Pull request closed without merging**